### PR TITLE
Add a unit test for BartModel to compare eager, sdpa on one particular set of inputs

### LINKED ISSUE #39365: Bug in modeling_bart.eager_attention_forward
### System Info

In [sdpa_forward_attention](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/sdpa_attention.py#L44),
we have this:

```python
if attention_mask is not None and attention_mask.ndim == 4:
    attention_mask = attention_mask[:, :, :, : key.shape[-2]]
```

In [eager_attention_foward](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bart/modeling_bart.py#L133), these two lines are not there and it may fail there:

```python
    if attention_mask is not None:
        attn_weights = attn_weights + attention_mask
```

I was wondering if that was intended. sdpa seems to be the default implementation.

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Switch default attn_implementation to eager.

### Expected behavior

Both implementation should work.

### PR DESCRIPTION
**Pull request by xadupre**
# What does this PR do?

Fixes #39365. Not a fix yet but introducing a unit test failing for the reason explained in this issue. Either the inputs are wrong, either the fix from issue #39365 is needed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/models/bart/test_modeling_bart.py [hunk 1]
```diff
@@ -316,6 +319,57 @@ def test_lm_uneven_forward(self):
...
+            encoder_ffn_dim=8,
+            decoder_ffn_dim=8,
+            max_position_embeddings=48,
+        )
+        config._attn_implementation = attn_implementation
+        lm_model = BartForConditionalGeneration(config).to(torch_device)
+        context = torch.tensor(
+            [[71, 82, 18, 33, 46, 91, 2], [68, 34, 26, 58, 30, 2, 1]], device=torch_device, dtype=torch.long
+        )
+        mask = torch.ones((context.shape[0], context.shape[1] + 2), device=context.device, dtype=torch.int64)
```
Comment (vasqu) [inline tests/models/bart/test_modeling_bart.py:340]: So it fails when we give it a mask that is too long for the original input?

File: tests/models/bart/test_modeling_bart.py [hunk 2]
```diff
@@ -316,6 +319,57 @@ def test_lm_uneven_forward(self):
...
+        shape1 = (2, 2, 7, 7)
+        shape2 = (2, 2, 2, 7)
+        past_key_values = EncoderDecoderCache(
```
Comment (vasqu) [inline tests/models/bart/test_modeling_bart.py:343]: I'd rather if we could just pass `use_cache` in the forward call.

File: tests/models/bart/test_modeling_bart.py [hunk 3]
```diff
@@ -316,6 +319,57 @@ def test_lm_uneven_forward(self):
         expected_shape = (*summary.shape, config.vocab_size)
         self.assertEqual(outputs["logits"].shape, expected_shape)
 
+    @parameterized.expand(["sdpa", "eager"])
+    def test_lm_uneven_forward_with_mask(self, attn_implementation):
```
Comment (vasqu) [inline tests/models/bart/test_modeling_bart.py:323]: I don't think we need to parametrize this; just set the config accordingly and check one after the other :eyes: 

Other than that, I think it makes sense to move this test under `BartModelTester`/`BartModelTest`, you can look at `test_encoder_decoder_model_standalone` for reference. We should possibly only modify the inputs (attention mask) and keep the rest from the defaults, i.e. `config_and_inputs = self.model_tester.prepare_config_and_inputs_for_common()`.

**Review by vasqu — COMMENTED** [review]
Some initial thoughts on the test: Does this only happen if we pass invalid masks? I.e. the mask was too long? 

Added some comments on the test design to make it more simple / reuse things we have.

**Comment by vasqu** [comment]
@xadupre Still any interest in this PR/bug? Otherwise, I'd make another PR to apply this fix and propagate it :eyes:

**Comment by qgallouedec** [comment]
Closing, see https://github.com/huggingface/transformers/issues/39365#issuecomment-5625523643
